README.md
# twoddpcr

[twoddpcr](https://bioconductor.org/packages/twoddpcr/) is an R/Bioconductor 
package for Droplet Digital PCR (ddPCR) analysis. The package is named for its 
ability to classify two channel ddPCR data (the amalgamation of "2-d" and 
"ddPCR").
                
Example usage of the package with a sample dataset is included in the package's 
[vignette](https://bioconductor.org/packages/release/bioc/vignettes/twoddpcr/inst/doc/twoddpcr.html), 
where the Appendix also explains how to extract raw droplet data from Bio-Rad's 
QuantaSoft.


## Installing the `twoddpcr` package

The package can be installed from Bioconductor using:

```
source("https://bioconductor.org/biocLite.R")
biocLite("twoddpcr")
```

Alternatively, it can be installed from GitHub using:

```
library(devtools)
install_github("CRUKMI-ComputationalBiology/twoddpcr")
```

Another alternative is to install the package from source:

```
install.packages("</path/to/twoddpcr/>", repos=NULL, type="source")
```


## Loading the `twoddpcr` package

Once the package has been installed, it can be loaded in the usual way:

```
library(twoddpcr)
```


## `ddpcrPlate` objects

The package revolves around the use of `ddpcrPlate` objects, which stores 
droplet amplitude data for the wells in the plate. The `ddpcrPlate` object also 
keeps track of the various classifications of the droplets.

To use your own droplet data and store it in a variable named `plate`, use:

```
plate <- ddpcrPlate("amplitudes/")
```

Here, we have assumed that your droplet amplitude CSVs are stored in the 
`amplitudes` directory inside the working directory.


## Classifying using the k-means clustering

If your data has four distinct clusters, the k-means clustering algorithm can 
be used:

```
plate <- kmeansClassify(plate)
```

This method can also take the `centres` parameter as a matrix to adjust the 
approximate centres of each cluster. This parameter can also be the number of 
clusters instead of a matrix. See `?kmeansClassify` for more details.


## Plotting classifications

Once the droplets have been classified, they can be plotted using the 
`dropletPlot` method. To use this, the names of the available classification 
methods can be obtained by:

```
getCommonClassMethod(plate)
```

There should be a `kmeans` entry for k-means clustering. To plot this, use:

```
dropletPlot(plate, cMethod="kmeans")
```


## Adding "rain"

Droplets between the main clusters of droplets will possibly have ambiguous 
classifications. These droplets are colloquially known as "rain". To do this, 
use:

```
plate <- mahalanobisRain(plate, maxDistances=list(NN=35, NP=30, PN=30, PP=30))
```

Some trial-and-error is required for setting the `maxDistances` parameter.


## Summaries

Since ddPCR samples are partitioned into droplets, the molecules are assumed to 
be Poisson distributed. With this, the actual number of starting fragments can 
be estimated:

```
plateSummary(plate, cMethod="kmeansMahRain")
```


## Shiny-based GUI for non-R users

A Shiny app is included in the package, which provides a GUI that allows 
interactive use of the package for ddPCR analysis. This can be run from an 
interactive R session using:

```
shinyVisApp()
```

It can also be accessed at 
[http://shiny.cruk.manchester.ac.uk/twoddpcr/](http://shiny.cruk.manchester.ac.uk/twoddpcr/).

If you wish to run the app from your own server, instructions are given in the 
[vignette](https://bioconductor.org/packages/devel/bioc/vignettes/twoddpcr/inst/doc/twoddpcr.html#shiny-based-gui-for-non-r-users).


# Citing `twoddpcr`

If you use the `twoddpcr` package in your work, please cite the [Bioinformatics 
paper](http://dx.doi.org/10.1093/bioinformatics/btx308). The citation can be 
obtained within R using:

```
citation("twoddpcr")
```



inst/NEWS
twoddpcr v1.4.1 (Release date: 2018-05-21)
============

Update contact details of maintainer.


twoddpcr v1.3.1 (Devel release date: 2018-01-10)
============

Bug fixes:
* Fix Issue #2: The 'plotLimits' parameter now works as intended for the
  'facetPlot' function when used with the 'cMethod' parameter.


twoddpcr v1.1.3 (Devel release date: 2017-07-31)
============

New features:
* Unit tests for facetPlot.


twoddpcr v1.1.2 (Devel release date: 2017-07-28)
============

New features:
* facetPlot now has a new 'cMethod' parameter for plotting the classifications 
  against each other.


twoddpcr v1.1.1 (Devel release date: 2017-07-17)
============

Bug fixes:
* Merge all bug fixes below.


twoddpcr v1.0.6 (Release date: 2017-06-08)
============

Bug fixes:
* Fix Issue #1 further: Catch all errors from uploading droplet amplitude CSVs.


twoddpcr v1.0.5 (Release date: 2017-06-01)
============

Bug fixes:
* Fix Issue #1 further: Stop the Shiny app from crashing when one channel data 
  or data in the wrong format is uploaded. Emphasis throughout that only two
  channel data is accepted.
* Make all hyperlinks open in a new tab/page. This prevents the user from 
  accidentally ending a session in the app.


twoddpcr v1.0.4 (Release date: 2017-05-30)
============

Bug fixes:
* Fix Issue #1: Loading a QuantaSoft classification no longer looks for a 
  column named "Cluster".

twoddpcr v1.0.3 (Release date: 2017-05-26)
============

Bug fixes:
* Ensure that the Rmd template is copied for report generation in the Shiny 
  app.

twoddpcr v1.0.2 (Release date: 2017-05-05)
============

Documentation:
* Update the citation information for the Bioinformatics paper.

twoddpcr v1.0.1 (Release date: 2017-05-04)
============

Documentation:
* In the Shiny app, add a reference to the Bioconductor vignette comparing the
  classification methods.

Bug fixes:
* Minor fix removing conflicts in the dropletPlot documentation.

twoddpcr v1.0.0 (Release date: 2017-04-25)
============

First full release incorporating all of the devel updates.

twoddpcr v0.99.2 (Release date: 2017-03-21)
============

Updates as suggested by anonymous reviewers.

Shiny app:
* Simple dynamic HTML report generation.
* Re-use the user-specified labels in plots and tables (where possible). NN, 
  NP, PN and PP are still used internally and for file input/output.
* Add captions to pages, improve the help sidebars and page.

Vignette:
* Describe the KRASdata dataset.
* Compare the classification methods with discussion.

Bug fixes:
* k-means failed when only one class was present.

twoddpcr v0.99.1 (Release date: 2017-02-16)
============

More R-like changes as suggested by Martin Morgan at Bioconductor.

Main changes:

* ddpcrPlate objects can be subsetted like a list
* The amplitudes() function replaces getDroplets() and getAllDroplets()
* wellClassification() and plateClassification() replace the old get/set
  methods and can be set using <-
* Changes to improve the robustness and clarity of the code


twoddpcr v0.99.0 (Release date: 2017-01-24)
============

First devel release.

twoddpcr is an R package for Droplet Digitial PCR (ddPCR) analysis. The name of 
the package reflects the package's support for analysing two channel ddPCR 
data.

Key functionality:

* Create ddpcrPlate objects
* kmeansClassify() classifies droplets using k-means clustering
* mahalanobisRain() refines the classification, removing ambiguous droplets


R/mahRain.R
#' @include global.R
#' @include themes.R
#' @import methods
#' @import stats
NULL

#' Mahalanobis distance of droplets from a distribution.
#'
#' Find the Mahalanobis distance of all droplets from a given distribution.
#'
#' @param droplets A data frame of droplets with \code{Ch1.Amplitude} and 
#' \code{Ch2.Amplitude} columns.
#' @param clusStats A list of statistics for a cluster generated by 
#' \code{\link{classStats}}. This should have a \code{mean}, \code{cov} and 
#' \code{cov.inv} values.
#'
#' @return A vector of numbers, where each one corresponds to the distance of 
#' that droplet from the given distribution.
#'
#' @author Anthony Chiu, \email{anthony.chiu@cruk.manchester.ac.uk}

.mahDist <- function(droplets, clusStats)
{
  if(is.null(clusStats$cov.inv))
    return(0)
  else
  {
    n <- seq_along(nrow(droplets))
    x <- split(droplets[, c("Ch1.Amplitude", "Ch2.Amplitude")], n)
    mapply(mahalanobis, x,
           MoreArgs=list(center=clusStats$mean, cov=clusStats$cov.inv, 
                         inverted=TRUE))
  }
}

#' Fuzzy clusters by bivariate normal distributions.
#'
#' Assume that the class \code{cl} is bivariate normally distributed. This 
#' method adds fuzziness to the class. Other classes are not changed.
#'
#' @param droplets A data frame of droplets with \code{Ch1.Amplitude} and 
#' \code{Ch2.Amplitude} columns, as well as a class column (see the parameter 
#' \code{classCol}).
#' @param cl The class to focus on. This should be either "NN", "NP", "PN" or 
#' "PP".
#' @param maxDistance An integer corresponding to the maximum Mahalanobis 
#' distance for which we will consider points to be members of the class, i.e. 
#' this is the level outside of which we consider droplets to be too far from 
#' the cluster.
#' @param classCol The column (name or number) from \code{droplets} 
#' representing the class.
#'
#' @return A factor corresponding to the class column with \code{Rain} entries 
#' added for the class \code{cl}.
#'
#' @author Anthony Chiu, \email{anthony.chiu@cruk.manchester.ac.uk}
#'
#' @examples
#' ## Use one of the wells in KRASdata.
#' aWell <- KRASdata[["E03"]]
#'
#' ## For all "NN" droplets, remove those Mahalanobis distance 10 away from the
#' ## mean of the "NN" cluster.
#' .classwiseMahalanobisRain(aWell, cl="NN", maxDistance=30,
#'                           classCol="Cluster")

.classwiseMahalanobisRain <- function(droplets, cl, maxDistance=30, 
                                      classCol="class")
{
  clusStats <- classStats(droplets, classCol=classCol)[[cl]]

  # Retain the droplets within the chosen maxDistance of the mean of its class.
  droplets[droplets[, classCol] == cl, classCol] <-
    ifelse(
      .mahDist(droplets[droplets[, classCol] == cl, ],
               clusStats=clusStats) <= maxDistance,
      as.character(droplets[droplets[, classCol] == cl, classCol]),
      ddpcr$rain)
  
  factor(droplets[, classCol], levels=ddpcr$classesRain)
}


#' Define 'rain' (unclassified) droplets by fitting the clusters to
#' bivariate normal distributions.
#'
#' Assume that each of the classified clusters are bivariate normally
#' distributed. We add fuzziness to the classifications by assigning droplets 
#' far away from the centres as "Rain". We use the Mahalanobis distance for 
#' each cluster to determine whether a droplet is 'too far away'.
#'
#' @param droplets A \code{ddpcrWell} or \code{ddpcrPlate} object, or 
#' a data frame of droplets with "Ch1.Amplitude" and "Ch2.Amplitude" columns, 
#' as well as a class column.
#' @param cMethod The name or column number of the classification for which we 
#' want to add rain to.
#' @param maxDistances A list of (levels) with keys in \code{c("NN", "PN", 
#' "NP", "PP")}. If the list is empty or set as \code{NULL}, no rain is added. 
#' If set as a single integer \code{n}, this value is taken for all classes, 
#' i.e. \code{list("NN"=n, "PN"=n, "NP"=n, "PP"=n)}.
#' @param ... Other options depending on the type of \code{droplets}.
#'
#' @return An object where the specified class has "Rain" entries added.
#'
#' @name mahalanobisRain
#'
#' @author Anthony Chiu, \email{anthony.chiu@cruk.manchester.ac.uk}
#'
#' @aliases multivariateRain multivariateNormalRain mvNormalRain mvnRain
#'
#' @examples
#' ## Take a data frame of droplets of transform it into the rigth format.
#' droplets <- KRASdata[["E03"]]
#' droplets$Cluster <- relabelClasses(droplets, classCol="Cluster")
#'
#' ## Add rain as a new column.
#' droplets$ClusterMahRain <-
#'     mahalanobisRain(droplets, cMethod="Cluster", fullTable=FALSE)
#' table(droplets$ClusterMahRain)
#'
#' ## The maximum distance around each mean can be changed uniformly.
#' droplets$ClusterMahRain <-
#'     mahalanobisRain(droplets, cMethod="Cluster", maxDistances=35,
#'                     fullTable=FALSE)
#' table(droplets$ClusterMahRain)
#'
#' ## Or we can change the maximum distances for each individual cluster.
#' droplets$ClusterMahRain <-
#'     mahalanobisRain(droplets, cMethod="Cluster",
#'                     maxDistances=list(NN=35, NP=30, PN=30, PP=30),
#'                     fullTable=FALSE)
#' table(droplets$ClusterMahRain)
#'
#' # This method works the same for ddpcrWell objects.
#' aWell <- ddpcrWell(well=KRASdata[["E03"]])
#' aWell <- mahalanobisRain(aWell, cMethod="Cluster")
#' table(wellClassification(aWell, cMethod="ClusterMahRain"))
#'
#' # Likewise for ddpcrPlate objects.
#' krasPlate <- ddpcrPlate(wells=KRASdata[c("E03", "H03", "C04", "F04")])
#' krasPlate <- mahalanobisRain(krasPlate, cMethod="Cluster")
#' lapply(plateClassification(krasPlate, cMethod="ClusterMahRain"), table)
#'
#' @export

setGeneric("mahalanobisRain", function(droplets, cMethod, maxDistances=30, ...)
  {
    standardGeneric("mahalanobisRain")
  }
)


#' @rdname mahalanobisRain
#'
#' @param fullTable If \code{TRUE}, returns a full data frame of droplets and 
#' their classification; if \code{FALSE}, simply returns a factor corresponding 
#' to this classification. Defaults to \code{TRUE}.
#'
#' @exportMethod mahalanobisRain

setMethod("mahalanobisRain", "data.frame",
  function(droplets, cMethod, maxDistances=30, fullTable=TRUE)
  {
    # maxDistances is just a number---assume that all of the classes 
    # have the same maximum distances
    if(is.numeric(maxDistances) && length(maxDistances) == 1)
    {
      n <- length(ddpcr$classes)
      maxDistances <- setNames(rep(maxDistances, n), ddpcr$classes)
    }
    # Assume that an empty vector means no rain.
    else if(is.null(maxDistances))
      maxDistances <- list()
    # Not a list and none of the conditions above---do not know how to 
    # handle.
    else if(!is.list(maxDistances))
      stop("'maxDistances' should be a named list with names in c(",
           ddpcr$nn, ", ", ddpcr$np, ", ", ddpcr$pn, ", ",
           ddpcr$pp, ")")

    # Loop through the maxDistances and check where droplets lie.
    for(className in names(maxDistances))
    {
      droplets[, cMethod] <-
        .classwiseMahalanobisRain(droplets, className,
                                  maxDistances[[className]], 
                                  classCol=cMethod)
                                 
    }
    
    if(fullTable)
      droplets[, c("Ch1.Amplitude", "Ch2.Amplitude", cMethod)]
    else
      droplets[, cMethod]
  }
)


#' @rdname mahalanobisRain
#'
#' @exportMethod mahalanobisRain

setMethod("mahalanobisRain", "ddpcrWell",
  function(droplets, cMethod, maxDistances=30)
  {
    cl <- wellClassification(droplets, cMethod=cMethod, withAmplitudes=TRUE)
    cl <- mahalanobisRain(cl, cMethod=cMethod, maxDistances=maxDistances,
                          fullTable=FALSE)
    wellClassification(droplets, paste0(cMethod, "MahRain")) <- cl
    droplets
  }
)


#' @rdname mahalanobisRain
#'
#' @exportMethod mahalanobisRain

setMethod("mahalanobisRain", "ddpcrPlate",
  function(droplets, cMethod, maxDistances=30)
  {
    cl <- plateClassification(droplets, cMethod=cMethod, withAmplitudes=TRUE)
    cl <- do.call(rbind, cl)
    rainy <- mahalanobisRain(cl, cMethod=cMethod, maxDistances=maxDistances, 
                             fullTable=FALSE)
    plateClassification(droplets, paste0(cMethod, "MahRain")) <- rainy
    droplets
  }
)

